### ISSUE TITLE
`SimpleSmt.insert_leaf(..)` return previous value

### ISSUE THREAD

**Issue by frisitano** [0xPolygonMiden/crypto#144]
Currently the return type of `SimpleSmt.insert_leaf(..)` is `Result<(), MerkleError>`.  I wonder if we should change this such that it returns the old value if it existed.  Additionally, to account for the case where no leaf existed we can return `None`.  The signature would look like:

```rust
pub fn insert_leaf(&mut self, key: u64, value: Word) -> Result<Option<Word>, MerkleError>
```

https://github.com/0xPolygonMiden/crypto/blob/52db23cd423e01a92eaf92abfd25aeef5b882c92/src/merkle/simple_smt/mod.rs#L205-L221

**Comment by bobbinth** [comment]
In the context of an SMT, all leaves are considered to be populated. So, technically all leaves exist - so, if a node does not exist in the compact data structure, we should probably return just `[ZERO; 4]`.

**Comment by bobbinth** [comment]
Closed by #149.
